### PR TITLE
V2Wizard: Review step 

### PR DESCRIPTION
**Pull request by mgold1234**
This adds the Review step to the V2Wizard.

### CODE REVIEW

File: src/Components/CreateImageWizardV2/CreateImageWizard.tsx [hunk 1]
```diff
@@ -197,7 +198,9 @@ const CreateImageWizard = () => {
             name="Review"
             id="step-review"
             footer={<CustomWizardFooter disableNext={true} />}
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/CreateImageWizard.tsx:200]: The "Next" button should be "Create image" for this step, but I'm not sure if that would in the scope of the Review step PR as I can see it being a part of a PR that will implement the form submit functionality :thinking:

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+  const source = useAppSelector((state) => selectAwsSource(state));
+  const awsAccountId = useAppSelector((state) => selectAwsAccountId(state));
+  const awsShareMethod = useAppSelector((state) => selectAwsShareMethod(state));
+
+  const { data, isError } = useGetSourceUploadInfoQuery(
+    {
+      id: parseInt(source?.id as string),
+    },
+    { skip: source === undefined }
+  );
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:96]: I can't see the data or isError being used anywhere in the component. Are those needed?

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+        {/*    <TextListItem component={TextListItemVariants.dt}>
+          {googleAccType?.[getState()?.values?.['google-account-type']] ===
+          'Domain'
+            ? 'Domain'
+            : 'Principal'}
+          </TextListItem> */}
+        {/*   <TextListItem component={TextListItemVariants.dd}>
+          {getState()?.values?.['google-email'] ||
+            getState()?.values?.['google-domain']}
+        </TextListItem>*/}
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:176]: These values can be found in the state:

```typescript
  gcp: {
    shareMethod: GcpShareMethod;
    accountType: GcpAccountType;
    email: string | undefined;
  };
```

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+                  <br />
+                </TextListItem>
+              )}
+              {getState()?.values?.['register-system'] ===
+                'register-now-rhc' && (
+                <TextListItem>
+                  Use remote host configuration (rhc) utility
+                  <br />
+                </TextListItem>
+              )}
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:526]:
```typescript
  registration: {
    registrationType: string;
    activationKey: ActivationKeys['name'];
  };
```

registrationType value from the store should be useful here.

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+                    <br />
+                    If using an activation key with command line registration,
+                    you must provide your organization&apos;s ID. Your
+                    organization&apos;s ID is{' '}
+                    {/*    {getState()?.values?.['subscription-organization-id'] !==
+                    undefined ? (
+                      getState()?.values?.['subscription-organization-id']
+                    ) : (
+                      <Spinner size="md" />
+                    )}*/}
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:548]: We don't store the organization-id anywhere yet. Either we could get it via an API request or we can also add it to the store and initialize it on the Registration step if needed.

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+          Reference ID:
+        </TextListItem>
+        <TextListItem component={TextListItemVariants.dd}>
+          {oscapProfile}
+        </TextListItem>
+      </TextList>
+      <br />
+    </TextContent>
+  );
+};*/
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:672]: The profile can be found in the store and all of the other information should be fetchable by an API request.

File: src/Components/CreateImageWizardV2/steps/Review/index.tsx [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                    toggleContent={'OpenSCAP'}
+                    onToggle={(_event, isExpandedOscapDetail) =>
+                        onToggleOscapDetails(isExpandedOscapDetail)
+                    }
+                    isExpanded={isExpandedOscapDetail}
+                    isIndented
+                    data-testid="oscap-detail-expandable"
+                >
+                    <OscapList />
+                </ExpandableSection>
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/index.tsx:198]: The Details PR was merged so this can be probably enabled :)

File: src/store/helpers.ts [hunk 8]
```diff
@@ -1,4 +1,5 @@
 import type { V1ListSourceResponse } from './provisioningApi';
 
-export const extractProvisioningList = (list: V1ListSourceResponse) =>
-  list?.data;
+export const extractProvisioningList = (
+  list: V1ListSourceResponse | undefined
+) => list?.data;
```
Comment (regexowl) [inline src/store/helpers.ts:5]: Why is this change needed? The undefined option seems to be included in the original API schema already.

```typescript
(alias) type V1ListSourceResponse = {
    data?: {
        id?: string | undefined;
        name?: string | undefined;
        source_type_id?: string | undefined;
        uid?: string | undefined;
    }[] | undefined;
```

**Review by regexowl — COMMENTED** [review]
Gave the PR a promised initial review and added some comments.There are also some review step tests in the `CreateImageWizard.test.tsx` test suite that could be re-enabled again :slightly_smiling_face:

File: src/store/helpers.ts (hunk 8 above)
Comment (mgold1234) [reply src/store/helpers.ts:5]: because I am using this in ReviewStepTextList- const { data: rawAzureSources, isSuccess: isSuccessAzureSources } =
    useGetSourceListQuery({ provider: 'azure' });
  const azureSources = extractProvisioningList(rawAzureSources); 
  and its complain on - 
  TS2345: Argument of type 'V1ListSourceResponse | undefined' is not assignable to parameter of type 'V1ListSourceResponse'.   Type 'undefined' is not assignable to type 'V1ListSourceResponse'

File: src/Components/CreateImageWizardV2/steps/Review/index.tsx [hunk 9]
```diff
@@ -0,0 +1,204 @@
+import React, { useEffect, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/index.tsx:1]: I think its good idea

File: src/Components/CreateImageWizardV2/CreateImageWizard.tsx (hunk 1 above)
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/CreateImageWizard.tsx:215]: I will wait to to submit funcionallty

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 5 above)
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:548]: what do you think is better one?

**Review by regexowl — COMMENTED** [review]
The Review step looks really nice! I've added some comments, let me know if something's unclear :)

**Comment by regexowl** [comment]
Also the commit structure in this PR is great!! :sparkles: The only thing I've noticed is that the "enable tests" commit https://github.com/RedHatInsights/image-builder-frontend/pull/1587/commits/59a0ed653d234d5c03c88a2c8e1906d2fdd6cf2e also contains some changes outside of the tests, would it be possible to move the changes to a better place? Maybe to https://github.com/RedHatInsights/image-builder-frontend/pull/1587/commits/b5815cc5ec4ff049da29e854e79e981adf467d73 and keep only the test changes in the test commit?

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+    </div>
+  );
+};
+
+export const ImageOutputList = () => {
+  const distribution = useAppSelector((state) => selectDistribution(state));
+  const arch = useAppSelector((state) => selectArchitecture(state));
+  return (
+    <TextContent>
+      <ReleaseLifecycle />
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:57]: so its not related to that pr right?

[the remark this answers is not in the document]
Comment (regexowl) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:57]: I think we can keep it separate from this PR.

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+      </ExpandableSection>
+      <ExpandableSection
+        toggleContent={'File system configuration'}
+        onToggle={(_event, isExpandedFSC) => onToggleFSC(isExpandedFSC)}
+        isExpanded={isExpandedFSC}
+        isIndented
+        data-testid="file-system-configuration-expandable"
+      >
+        <FSCList />
+      </ExpandableSection>
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx:140]: amm not sure regarding that, it will be strange to remove this extandable form the review, maybe its good to leave it empty for now, @lucasgarfield wdyt?

[the remark this answers is not in the document]
Comment (regexowl) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx:140]: If it's too much commented code (which it probably is thinking about the toggles) then let's leave it. :)

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        onToggle={(_event, isExpandedTargetEnvs) =>
+          onToggleTargetEnvs(isExpandedTargetEnvs)
+        }
+        isExpanded={isExpandedTargetEnvs}
+        isIndented
+        data-testid="target-environments-expandable"
+      >
+        {environments.includes('aws') && <TargetEnvAWSList />}
+        {environments.includes('gcp') && <TargetEnvGCPList />}
+        {environments.includes('azure') && <TargetEnvAzureList />}
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx]: removed those lines

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+                aria-label="About activation key"
+                className="pf-u-pl-sm pf-u-pt-0 pf-u-pb-0"
+                size="sm"
+              >
+                <HelpIcon />
+              </Button>
+            </Popover>
+          </TextListItem>
+          <TextListItem component={TextListItemVariants.dd}>
+            <ActivationKeyInformation />
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:407]: this data comes from activationKey
that defined in the wizardSlice, do you think I need to change it in this pr?

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 14]
```diff
@@ -0,0 +1,495 @@
...
+        <TextListItem component={TextListItemVariants.dd}>
+          us-east-1
+        </TextListItem>
+      </TextList>
+      <br />
+    </TextContent>
+  );
+};
+
+export const TargetEnvGCPList = () => {
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:143]: Screenshot 2024-02-01 at 12 24 07
this is how its look in stage, not sure I understand what to change

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 15]
```diff
@@ -0,0 +1,495 @@
...
+};
+export const FSCList = () => {
+  return (
+    <TextContent>
+      <br />
+    </TextContent>
+  );
+};
+
+export const TargetEnvAWSList = () => {
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:88]: I think we should fix that in a different pr as you suggest

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+  const onToggleRegistration = (isExpandedRegistration: boolean) =>
+    setIsExpandedRegistration(isExpandedRegistration);
+  const onToggleImageDetail = (isExpandedImageDetail: boolean) =>
+    setIsExpandedImageDetail(isExpandedImageDetail);
+  const onToggleOscapDetails = (isExpandedOscapDetail: boolean) =>
+    setIsExpandedOscapDetail(isExpandedOscapDetail);
+
+  return (
+    <>
+      <ExpandableSection
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx:71]: good idea thank you @regexowl for your amazing review

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (regexowl) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:142]: For GCP step we currently have two options: "Share image with a Google acount" and "Share image with Red Hat Insights only". For the "with a Google Account" it makes perfect sense to show an account type and principal, but if we choose the "with Insights only" option, we shouldn't be able to show any account type or principal, because none were chosen. I think in case the image is shared only with insights, we should probably only show something like:

GCP
Image type / Red Hat hosted image
Shared with / Red Hat Insights only

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:402]: I think there is bug in activationKeyInformation - activationKey - is not consist any data.
I fixed that, tell me if it is correct

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:60]: fix it in https://github.com/RedHatInsights/image-builder-frontend/pull/1623

File: src/Components/CreateImageWizardV2/CreateImageWizard.tsx [hunk 17]
```diff
@@ -211,8 +212,10 @@ const CreateImageWizard = () => {
           <WizardStep
             name="Review"
             id="step-review"
-            footer={<CustomWizardFooter disableNext={true} />}
-          ></WizardStep>
+            footer={<CustomWizardFooter disableNext={false} />}
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/CreateImageWizard.tsx:215]: This is very much a nit pick, but let's keep the Next disabled for now so we can't jump out of the Wizard this way.

```suggestion
            footer={<CustomWizardFooter disableNext={true} />}
```

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (regexowl) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStep.tsx:72]: Looks much cleaner now, thank you! :heart:

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 18]
```diff
@@ -0,0 +1,532 @@
...
+  selectCustomRepositories,
+  selectDistribution,
+  selectGcpAccountType,
+  selectGcpEmail,
+  selectGcpShareMethod,
+  selectProfile,
+  selectRegistrationType,
+} from '../../../../store/wizardSlice';
+import { useGetEnvironment } from '../../../../Utilities/useGetEnvironment';
+import { RepositoriesTable } from '../../../CreateImageWizard/formComponents/ReviewStepTables';
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:45]: This links to the original Wizard so if the user clicks on the number of Custom repositories on the Review step the whole Wizard crashes with `getState is not a function`

image

```suggestion
import { RepositoriesTable } from './ReviewStepTables';
```

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 19]
```diff
@@ -0,0 +1,532 @@
...
+                aria-label="About activation key"
+                className="pf-u-pl-sm pf-u-pt-0 pf-u-pb-0"
+                size="sm"
+              >
+                <HelpIcon />
+              </Button>
+            </Popover>
+          </TextListItem>
+          <TextListItem component={TextListItemVariants.dd}>
+            <ActivationKeyInformation />
```
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:420]: The activation key on the Review step still shows the "Selected activation key" heading. Similar to the ReleaseLifecycle we might want to solve this one in a separate PR.

Thinking about this I can see two approaches to this:

First approach is we create a description list with the needed values directly in this file. That would mean copying a long blob of `<TextContent>` code over from `ActivationKeyInformation` which would mean a lot of duplicity. Not ideal, would get rid of the `FormGroup` label.

Second option would be that we create a separate PR and do the decisions directly in the `ActivationKeyInformation` component (I would much prefer this solution). We can figure out on which step we currently are in the Wizard: https://www.patternfly.org/components/wizard#get-current-step and based on this information we can render / not render the `FormGroup`. So it would go roughly like this:
- in the ActivationKeyInformation
- get currentStep of the Wizard
- if currentStep === 'registration' render FormGroup else render just TextContent

File: src/Components/CreateImageWizardV2/steps/Registration/ActivationKeyInformation.tsx [hunk 20]
```diff
@@ -21,7 +21,7 @@ import { useShowActivationKeyQuery } from '../../../../store/rhsmApi';
 import { selectActivationKey } from '../../../../store/wizardSlice';
 
 const ActivationKeyInformation = (): JSX.Element => {
-  const activationKey = useAppSelector((state) => selectActivationKey(state));
+  let activationKey = useAppSelector((state) => selectActivationKey(state));
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Registration/ActivationKeyInformation.tsx:24]: sure, I will do it in a seperate pr

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 18 above)
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:45]: woo stupit mistake

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 19 above)
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:420]: I agree, should be at a seperate pr

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 21]
```diff
@@ -0,0 +1,532 @@
...
+          <br />
+          <ExpirationWarning />
+        </TextListItem>
+        <TextListItem component={TextListItemVariants.dt}>
+          Shared to account
+        </TextListItem>
+        <TextListItem component={TextListItemVariants.dd}>
+          {!isBeta() && awsAccountId}
+          {isBeta() && awsShareMethod === 'sources' && isSuccess && (
+            <AwsAccountId />
```
Comment (lucasgarfield) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:119]: Let's replace `<AwsAccountId />` with `{awsAccountId}`.

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 22]
```diff
@@ -0,0 +1,532 @@
...
+        <TextListItem component={TextListItemVariants.dd}>
+          Red Hat hosted image
+          <br />
+          <ExpirationWarning />
+        </TextListItem>
+        <TextListItem component={TextListItemVariants.dt}>
+          Shared to account
+        </TextListItem>
+        <TextListItem component={TextListItemVariants.dd}>
+          {!isBeta() && awsAccountId}
```
Comment (lucasgarfield) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:117]: I think that now that Provisioning is GA we can get rid of the `isBeta()` checks... so:
```
        <TextListItem component={TextListItemVariants.dd}>
          {awsShareMethod === 'sources' && isSuccess && awsAccountId}
          {awsShareMethod === 'manual' && awsAccountId}
        </TextListItem>
```

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx [hunk 23]
```diff
@@ -0,0 +1,532 @@
...
+    </TextContent>
+  );
+};
+
+export const TargetEnvAWSList = () => {
+  const { data: rawAWSSources, isSuccess } = useGetSourceListQuery({
+    provider: 'aws',
+  });
+  const awsAccountId = useAppSelector((state) => selectAwsAccountId(state));
+  const awsShareMethod = useAppSelector((state) => selectAwsShareMethod(state));
```
Comment (lucasgarfield) [inline src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:93]: I think we need to add a selector to get the awsSource here if it exists.

Comment (lucasgarfield) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:93]: Let's do a follow-up PR.

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 22 above)
Comment (lucasgarfield) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:117]: Let's do a follow-up PR.

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 21 above)
Comment (lucasgarfield) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:119]: Let's do a follow-up PR.

File: src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx (hunk 19 above)
Comment (mgold1234) [reply src/Components/CreateImageWizardV2/steps/Review/ReviewStepTextLists.tsx:420]: opened this bug for that -  https://issues.redhat.com/browse/HMS-2781
